[19.05] Run futurize for python2 cheetah templates, fix scope issue for py3 comprehensions

mvdbeek opened the pull request:
I think this should cover the python 2/3 errors I have seen so far

mvdbeek committed:
Make variables that have not been found global

This will allow tools to work that access variables from
the enclosing scope in a list/dict/tuple comprehension.

mvdbeek committed:
Add test that unfortunately shows that the global variable isn't updated

Now if we could figure out the exact position in the template source
where a NotFound error is raised we could replace `$variable`
with `variable` and add `#variable = $variable` just before the failing
statement.

mvdbeek committed:
Run futurize on Compiled Cheetah Module

mvdbeek committed:
Work around list comprehension scope issues

mvdbeek commented:
Hah, turns out dict, set and generator comprehensions all have that problem on python 2, so I guess we don't need to fix that.

mvdbeek committed:
Non-list comprehensions can't refer to non-iteration variables from surrounding scope

While fixable this seems to not work on python 2 either, so we will not
attempt to fix it.

nsoranzo commented on lib/galaxy/util/template.py:
```diff
@@ -15,4 +49,60 @@ def fill_template(template_text, context=None, **kwargs):
+    except Exception as e:
+        if first_exception is None:
+            first_exception = e
+        if not futurized:
```
```suggestion
        if not futurized and sys.version_info.major > 2:
```

mvdbeek replied:
Thanks, I've added that to https://github.com/galaxyproject/galaxy/pull/7867/commits/8f692cbebd14bf4108ed050867aeb60bc54425d5

mvdbeek committed:
Only run python 3 fixers if profile version < 19.05

or python_template_version < 3.

nsoranzo commented on test/unit/tools/test_evaluation.py:
```diff
@@ -266,6 +266,7 @@ class MockTool(object):
 
     def __init__(self, app):
         self.profile = 16.01
+        self.python_template_version = 2.7
```
```suggestion
        self.python_template_version = '2.7'
```

nsoranzo replied:
```suggestion
        self.python_template_version = '2.7'
```

nsoranzo commented on lib/galaxy/tools/__init__.py:
```diff
@@ -643,6 +644,14 @@ def parse(self, tool_source, guid=None, dynamic=False):
+        if self.python_template_version is None:
+            # If python_template_version not specified we assume tools with profile versions >= 19.05 are python 3 ready
+            if self.profile >= 19.05:
+                self.python_template_version = 3.5
```
```suggestion
                self.python_template_version = '3.5'
```

nsoranzo commented on lib/galaxy/tools/__init__.py:
```diff
@@ -643,6 +644,14 @@ def parse(self, tool_source, guid=None, dynamic=False):
+            if self.profile >= 19.05:
+                self.python_template_version = 3.5
+            else:
+                self.python_template_version = 2.7
```
```suggestion
                self.python_template_version = '2.7'
```

nsoranzo commented on lib/galaxy/tools/actions/__init__.py:
```diff
@@ -925,7 +926,7 @@ def get_ext_or_implicit_ext(hda):
 
 
-def determine_output_format(output, parameter_context, input_datasets, input_dataset_collections, random_input_ext):
+def determine_output_format(output, parameter_context, input_datasets, input_dataset_collections, random_input_ext, python_template_version=3):
```
```suggestion
def determine_output_format(output, parameter_context, input_datasets, input_dataset_collections, random_input_ext, python_template_version='3'):
```

nsoranzo commented on lib/galaxy/tools/parser/cwl.py:
```diff
@@ -150,6 +150,9 @@ def parse_requirements_and_containers(self):
         return "16.04"
 
+    def parse_python_template_version(self):
+        return 3.5
```
```suggestion
        return '3.5'
```

nsoranzo commented on lib/galaxy/tools/parser/xml.py:
```diff
@@ -483,6 +483,12 @@ def parse_profile(self):
+    def parse_python_template_version(self):
+        python_template_version = self.root.get("python_template_version", None)
+        if python_template_version is not None:
+            python_template_version = float(python_template_version)
```
```suggestion
            python_template_version = packaging.version.parse(python_template_version)
```

You'd need to also add `import packaging.version` at the top.

nsoranzo commented on lib/galaxy/tools/parser/yaml.py:
```diff
@@ -162,6 +162,12 @@ def parse_tests_to_dict(self):
+    def parse_python_template_version(self):
+        python_template_version = self.root_dict.get("python_template_version", None)
+        if python_template_version is not None:
+            python_template_version = float(python_template_version)
```
```suggestion
            python_template_version = packaging.version.parse(python_template_version)
```

You'd need to also add `import packaging.version` at the top.

nsoranzo commented on lib/galaxy/util/template.py:
```diff
@@ -1,10 +1,45 @@
+                  compiler_class=Compiler,
+                  first_exception=None,
+                  futurized=False,
+                  python_template_version=3,
```
```suggestion
                  python_template_version='3',
```

nsoranzo commented on lib/galaxy/util/template.py:
```diff
@@ -15,4 +50,62 @@ def fill_template(template_text, context=None, **kwargs):
+                                     context=context,
+                                     retry=retry - 1,
+                                     compiler_class=compiler_class,
+                                     first_exception=first_exception
```
Add:
```python

                                     python_template_version=python_template_version,
```

nsoranzo commented on lib/galaxy/util/template.py:
```diff
@@ -15,4 +50,62 @@ def fill_template(template_text, context=None, **kwargs):
+                                 retry=retry,
+                                 compiler_class=compiler_class,
+                                 first_exception=first_exception,
+                                 futurized=True
```
Add:
```python

                                     python_template_version=python_template_version,
```

nsoranzo commented on lib/galaxy/util/template.py:
```diff
@@ -1,10 +1,45 @@
+import sys
+import tempfile
+import traceback
+
```
Add:
```python
import packaging.version
```

nsoranzo commented on lib/galaxy/util/template.py:
```diff
@@ -15,4 +50,62 @@ def fill_template(template_text, context=None, **kwargs):
         raise TypeError("Template text specified as None to fill_template.")
     if not context:
         context = kwargs
-    return unicodify(Template(source=template_text, searchList=[context]))
```
Add:
```python
    if isinstance(python_template_version, str):
        python_template_version = packaging.version.parse(python_template_version)
```

nsoranzo commented on lib/galaxy/util/template.py:
```diff
@@ -15,4 +50,62 @@ def fill_template(template_text, context=None, **kwargs):
+    except NotFound as e:
+        if first_exception is None:
+            first_exception = e
+        if python_template_version < 3 and retry > 0 and sys.version_info.major > 2:
```
```suggestion
        if sys.version_info.major > 2 and python_template_version.release[0] < 3 and retry > 0:
```

nsoranzo commented on lib/galaxy/util/template.py:
```diff
@@ -15,4 +50,62 @@ def fill_template(template_text, context=None, **kwargs):
+    except Exception as e:
+        if first_exception is None:
+            first_exception = e
+        if sys.version_info.major > 2 and python_template_version < 3 and not futurized:
```
```suggestion
        if sys.version_info.major > 2 and python_template_version.release[0] < 3 and not futurized:
```

nsoranzo commented on test/unit/tools/test_fill_template.py:
```diff
@@ -0,0 +1,77 @@
+
+
+def test_fill_list_comprehension_template_2():
+    template_str = fill_template(LIST_COMPREHENSION_TEMPLATE, python_template_version=2, retry=1)
```
```suggestion
    template_str = fill_template(LIST_COMPREHENSION_TEMPLATE, python_template_version='2', retry=1)
```

nsoranzo commented on test/unit/tools/test_fill_template.py:
```diff
@@ -0,0 +1,77 @@
+
+
+def test_fix_template_two_to_three():
+    template_str = fill_template(TWO_TO_THREE_TEMPLATE, python_template_version=2, retry=1)
```
```suggestion
    template_str = fill_template(TWO_TO_THREE_TEMPLATE, python_template_version='2', retry=1)
```

nsoranzo reviewed (commented):
I'd prefer to parse `python_template_version` with `packaging.version.parse()` instead of casting with `float()`, see comments inline.

nsoranzo committed:
Use packaging.version.parse() instead of `float()` for version comparisons

mvdbeek committed:
Use lib2to3 RefactoringTool instead of external process

We use a list of fixers from past.translation. This is just a
list of strings pointing to fixers to import. Currently this is
```
['libfuturize.fixes.fix_absolute_import',
 'libfuturize.fixes.fix_print_with_import',
 'libfuturize.fixes.fix_raise',
 'libfuturize.fixes.fix_next_call',
 'lib2to3.fixes.fix_reduce',
 'lib2to3.fixes.fix_sys_exc',
 'lib2to3.fixes.fix_throw',
 'lib2to3.fixes.fix_tuple_params',
 'lib2to3.fixes.fix_exec',
 'lib2to3.fixes.fix_standarderror',
 'lib2to3.fixes.fix_apply',
 'lib2to3.fixes.fix_paren',
 'lib2to3.fixes.fix_types',
 'lib2to3.fixes.fix_methodattrs',
 'lib2to3.fixes.fix_renames',
 'lib2to3.fixes.fix_ws_comma',
 'lib2to3.fixes.fix_repr',
 'lib2to3.fixes.fix_intern',
 'lib2to3.fixes.fix_numliterals',
 'lib2to3.fixes.fix_funcattrs',
 'lib2to3.fixes.fix_ne',
 'lib2to3.fixes.fix_exitfunc',
 'lib2to3.fixes.fix_has_key',
 'lib2to3.fixes.fix_isinstance',
 'lib2to3.fixes.fix_idioms',
 'lib2to3.fixes.fix_xreadlines',
 'lib2to3.fixes.fix_except',
 'libfuturize.fixes.fix_execfile',
 'libfuturize.fixes.fix_future_standard_library',
 'libfuturize.fixes.fix_metaclass',
 'libfuturize.fixes.fix_future_builtins',
 'libpasteurize.fixes.fix_newstyle',
 'libfuturize.fixes.fix_basestring',
 'libfuturize.fixes.fix_future_standard_library_urllib',
 'libfuturize.fixes.fix_xrange_with_import',
 'libfuturize.fixes.fix_cmp',
 'libfuturize.fixes.fix_unicode_keep_u',
 'libfuturize.fixes.fix_division_safe',
 'libfuturize.fixes.fix_object',
 'lib2to3.fixes.fix_input',
 'lib2to3.fixes.fix_operator',
 'lib2to3.fixes.fix_zip',
 'lib2to3.fixes.fix_dict',
 'lib2to3.fixes.fix_next',
 'lib2to3.fixes.fix_filter',
 'lib2to3.fixes.fix_raw_input',
 'lib2to3.fixes.fix_getcwdu',
 'lib2to3.fixes.fix_itertools_imports',
 'lib2to3.fixes.fix_nonzero',
 'lib2to3.fixes.fix_itertools',
 'lib2to3.fixes.fix_long',
 'lib2to3.fixes.fix_map']
```
And then we drop pasterurize, which we don't need for our purposes.

mvdbeek committed:
Delete running jobs after each API-style test

Should prevent https://jenkins.galaxyproject.org/job/docker-integration-py3/159/testReport/junit/test.integration.test_upload_configuration_options/AdminsCanPasteFilePathsTestCase/test_admin_path_paste_libraries/

mvdbeek committed:
Don't encode string, this is unnecessary and fails on python 3

mvdbeek commented:
@galaxybot test py3

martenson commented:
merged forward
